lib/net/ldap/connection.rb

```diff
@@ -8,56 +8,49 @@ class Net::LDAP::Connection #:nodoc:
 
   def initialize(server)
     @instrumentation_service = server[:instrumentation_service]
-    server[:hosts] = [[server[:host], server[:port]]] if server[:hosts].nil?
 
     if server[:socket]
-      prepare_socket(server)
+      prepare_socket(server[:socket], server[:encryption])
     else
-      open_connection(server)
+      hosts = server[:hosts]
+      hosts = [[server[:host], server[:port]]] if hosts.nil?
+      open_connection(hosts, server[:encryption])
     end
 
     yield self if block_given?
   end
 
-  def prepare_socket(server)
-    @conn = server[:socket]
-
-    if server[:encryption]
-      setup_encryption server[:encryption]
-    end
+  def prepare_socket(socket, encryption, close = false)
+    @conn = socket
+    setup_encryption encryption if encryption
+  rescue
+    # Ensure the connection is closed when requested in the event of an SSL
+    # setup failure.
+    @conn.close if close
+    @conn = nil
+    raise
   end
 
-  def open_connection(server)
+  def open_connection(hosts, encryption)
```

Comment:
This is mostly to make it explicit what arguments these methods actually use.  Passing around a hash makes it necessary to inspect the guts of the method to know what is really needed.